diff -ruN /hbase-2.2.4/hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java ./hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java
--- /hbase-2.2.4/hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java	2020-03-11 04:27:36.000000000 +0000
+++ ./hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java	2020-04-09 19:52:19.625704238 +0000
@@ -492,6 +492,7 @@
   public HMaster(final Configuration conf)
       throws IOException, KeeperException {
     super(conf);
+    LOG.info("msx-hbase HMaster init");
     TraceUtil.initTracer(conf);
     try {
       if (conf.getBoolean(MAINTENANCE_MODE, false)) {
@@ -571,7 +572,8 @@
   // Main run loop. Calls through to the regionserver run loop AFTER becoming active Master; will
   // block in here until then.
   @Override
-  public void run() {
+  public void run() { 
+    LOG.info("msx-hbase HMaster start");
     try {
       if (!conf.getBoolean("hbase.testing.nocluster", false)) {
         Threads.setDaemonThreadRunning(new Thread(() -> {
@@ -2785,6 +2787,7 @@
   }
 
   public void stopMaster() throws IOException {
+    LOG.info("msx-hbase HMaster stop");
     if (cpHost != null) {
       cpHost.preStopMaster();
     }
diff -ruN /hbase-2.2.4/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java ./hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
--- /hbase-2.2.4/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	2020-03-11 04:27:36.000000000 +0000
+++ ./hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java	2020-04-09 20:42:33.773621369 +0000
@@ -564,6 +564,7 @@
   // Defer till after we register with the Master as much as possible. See #startServices.
   public HRegionServer(Configuration conf) throws IOException {
     super("RegionServer");  // thread name
+    LOG.info("msx-hbase HRegionServer init");
     TraceUtil.initTracer(conf);
     try {
       this.startcode = System.currentTimeMillis();
@@ -969,6 +970,7 @@
    */
   @Override
   public void run() {
+    LOG.info("msx-hbase HRegionServer start");
     try {
       // Do pre-registration initializations; zookeeper, lease threads, etc.
       preRegistrationInitialization();
@@ -2217,6 +2219,7 @@
    */
   public void stop(final String msg, final boolean force, final User user) {
     if (!this.stopped) {
+      LOG.info("msx-hbase HRegionServer stop");
       LOG.info("***** STOPPING region server '" + this + "' *****");
       if (this.rsHost != null) {
         // when forced via abort don't allow CPs to override
