diff -ruN /hadoop-2.8.5-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java
--- /hadoop-2.8.5-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java	2018-09-09 08:35:07.000000000 +0000
+++ ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java	2020-06-04 03:11:47.461879671 +0000
@@ -104,6 +104,10 @@
 import com.google.common.base.Preconditions;
 import com.google.common.base.Strings;
 
+// msx
+import java.io.BufferedReader;
+import java.io.FileReader;
+
 /**
  * Provides access to configuration parameters.
  *
@@ -182,7 +186,7 @@
  */
 @InterfaceAudience.Public
 @InterfaceStability.Stable
-public class Configuration implements Iterable<Map.Entry<String,String>>,
+public class Configuration extends ReconfAgent implements Iterable<Map.Entry<String,String>>,
                                       Writable {
   private static final Log LOG =
     LogFactory.getLog(Configuration.class);
@@ -709,7 +713,10 @@
       handleDeprecation(deprecations, (String)item);
     }
   }
- 
+
+  // msx
+  private static boolean msxConfEnable = false;
+
   static{
     //print deprecation warning if hadoop-site.xml is found in classpath
     ClassLoader cL = Thread.currentThread().getContextClassLoader();
@@ -725,6 +732,18 @@
     }
     addDefaultResource("core-default.xml");
     addDefaultResource("core-site.xml");
+    // msx
+    try {
+        BufferedReader reader = new BufferedReader(new FileReader("/root/reconf_test_gen/lib/enable"));
+        String buffer = reader.readLine();
+        reader.close();
+        if (buffer.equals("true"))
+            msxConfEnable = true;
+        else
+            msxConfEnable = false;
+    } catch(Exception e) {
+        e.printStackTrace();
+    }
   }
   
   private Properties properties;
@@ -1071,7 +1090,17 @@
     throw new IllegalStateException("Variable substitution depth too large: " 
                                     + MAX_SUBST + " " + expr);
   }
-  
+ 
+  // msx
+  private void whoInvokesMe(String parameter, String returnValue) {
+      if (msxConfEnable == false)
+          return;
+      String getMethodName = Thread.currentThread().getStackTrace()[2].getMethodName();
+      System.out.println("msx-conf parameter " + parameter + " getMethod " + getMethodName + " conf-hashcode "
+          + this.hashCode() + " returnValue " + returnValue);
+      return;
+  }
+
   /**
    * Get the value of the <code>name</code> property, <code>null</code> if
    * no such property exists. If the key is deprecated, it returns the value of
diff -ruN /hadoop-2.8.5-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configured.java ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configured.java
--- /hadoop-2.8.5-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configured.java	2018-05-05 10:34:35.000000000 +0000
+++ ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configured.java	2020-06-03 22:32:41.358616706 +0000
@@ -24,7 +24,7 @@
 /** Base class for things that may be configured with a {@link Configuration}. */
 @InterfaceAudience.Public
 @InterfaceStability.Stable
-public class Configured implements Configurable {
+public class Configured extends ReconfAgent implements Configurable {
 
   private Configuration conf;
 
diff -ruN /hadoop-2.8.5-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/ReconfAgent.java ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/ReconfAgent.java
--- /hadoop-2.8.5-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/ReconfAgent.java	1970-01-01 00:00:00.000000000 +0000
+++ ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/ReconfAgent.java	2020-06-04 02:07:29.133812388 +0000
@@ -0,0 +1,211 @@
+package org.apache.hadoop.conf;
+
+import java.io.*;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.HashMap;
+import org.apache.hadoop.conf.Configuration;
+
+public class ReconfAgent {
+
+    //private static final int RP_MODE_INSTANCE = -1; // set v2 to a single instance of the specified component throughout a test
+    //private static final int RP_MODE_COMPONENT = -2; // set v2 to all instances for the specified component
+    //private static final String RP_MODE_RECONF = "1"; // set v2 to a single life cycle of a single instance of the specified component
+    private static final String reconf_systemRootDir = "/root/parameter_test_controller/";
+
+    private static String reconf_vvmode = "";
+    private static String reconf_parameter = "";
+    private static String reconf_component = "";
+    private static String reconf_v1 = "";
+    private static String reconf_v2 = "";
+    private static String reconf_point = "";
+    private static int reconf_point_int = 0; 
+    
+    private static int reconf_init_point_index = 0;
+    private static Map<Integer, String> confHcComponentMap = new HashMap<Integer, String>();
+    private static List<Object> componentList = new ArrayList<Object>();
+
+    // load just once
+    static {
+	loadSharedVariables();
+    }
+
+    private static void loadSharedVariables() {
+        try {
+            BufferedReader reader;
+            reader = new BufferedReader(new FileReader(new File(reconf_systemRootDir + "shared/reconf_vvmode")));
+            reconf_vvmode = reader.readLine();
+            reader.close();
+
+            reader = new BufferedReader(new FileReader(new File(reconf_systemRootDir + "shared/reconf_parameter")));
+            reconf_parameter = reader.readLine();
+            reader.close();
+
+            reader = new BufferedReader(new FileReader(new File(reconf_systemRootDir + "shared/reconf_component")));
+            reconf_component = reader.readLine();
+            reader.close();
+
+            reader = new BufferedReader(new FileReader(new File(reconf_systemRootDir + "shared/reconf_v1")));
+            reconf_v1 = reader.readLine();
+            reader.close();
+
+            reader = new BufferedReader(new FileReader(new File(reconf_systemRootDir + "shared/reconf_v2")));
+            reconf_v2 = reader.readLine();
+            reader.close();
+
+            reader = new BufferedReader(new FileReader(new File(reconf_systemRootDir + "shared/reconf_point")));
+            reconf_point = reader.readLine();
+            reader.close();
+            reconf_point_int = Integer.valueOf(reconf_point);
+
+            if (!reconf_vvmode.equals("v1v1") && !reconf_vvmode.equals("v2v2") && !reconf_vvmode.equals("v1v2") && !reconf_vvmode.equals("none")) {
+                myPrint("ERROR : wrong value of reconf_vvmode " + reconf_vvmode);
+                System.exit(1);
+            }
+      
+	    myPrint("reconf_vvmode=" + reconf_vvmode + ", reconf_parameter=" + reconf_parameter + 
+			    ", reconf_component=" + reconf_component + ", reconf_v1=" + reconf_v1 + ", reconf_v2=" + reconf_v2 +
+			    ", reconf_point=" + reconf_point);
+        } catch (Exception e) {
+            myPrint("ERROR : loadSharedVariables");
+            e.printStackTrace();
+        }
+    }
+
+    /*
+    public static void checkReconfAtShutdown(Object obj, String component, Configuration myConf) {
+	if (obj == null || myConf == null)
+	    return;
+        myPrint("" + component + " stop " + obj.hashCode() + ", value is " + myConf.get(reconf_parameter));
+	Integer confToRemove = new Integer(myConf.hashCode());
+	String v = confHcComponentMap.remove(confToRemove);
+	if (v == null) {
+	    myPrint("WARN : conf " + confToRemove + " not existed in confInstanceList when removing.");
+	}
+    }*/
+
+    public synchronized static void performReconf(Object obj, String component, Configuration myConf) {
+      if (myConf == null) {
+	  myPrint("ERROR : myConf is null");
+	  return;
+      }
+      String componentHcStr = "";
+      // check if component instance is already registered
+      if (obj != null) {
+          if (componentList.contains(obj)) {
+              myPrint("ERROR : component " + component + " hc " + Integer.toString(obj.hashCode()) + " already existed in componentList, just ignore this one.");
+              return;
+          } else {
+              componentList.add(obj);
+              componentHcStr = Integer.toString(obj.hashCode());
+          }
+      } else {
+          componentHcStr = "static";
+      }
+      
+      // check if component instance's conf is shared with other component instances
+      Integer confToAdd = new Integer(myConf.hashCode());
+      if (!confHcComponentMap.containsKey(confToAdd)) {
+	  confHcComponentMap.put(confToAdd, component);
+	  myPrint("conf with hashCode " + confToAdd + " for " + component);
+      } else {
+	  myPrint("WARN : conf " + confToAdd + " existed in confInstanceList when adding.");
+	  Configuration copiedConf = null;
+	  copiedConf = new Configuration(myConf);
+	  myConf = copiedConf;
+	  confToAdd = new Integer(myConf.hashCode());
+	  if (!confHcComponentMap.containsKey(confToAdd)) {
+	      confHcComponentMap.put(confToAdd, component);
+	      myPrint("conf with hashCode " + confToAdd + " for " + component);
+	      myPrint("issue is fixed with copied configuration.");
+	  } else {
+	      myPrint("ERROR : conf " + confToAdd + " existed in confInstanceList when adding.");
+	  }
+      }
+
+      myPrint("performReconf for comoponent_name " + component + " component_hc " + componentHcStr + " conf_hc " + confToAdd);
+
+      if (reconf_vvmode.equals("none")) {
+          myPrint(component + " init, vvmode is none, do nothing");
+      }
+
+      if (reconf_vvmode.equals("v1v1")) {
+          myConf.set(reconf_parameter, reconf_v1);
+          myPrint(component + " init " + componentHcStr + ", vvmode is " + reconf_vvmode + 
+			  ". Set value as v1 " + reconf_v1);// + " myConf is " + myConf.hashCode());
+      }
+
+      if ((reconf_vvmode.equals("v2v2"))) {
+          myConf.set(reconf_parameter, reconf_v2);
+          myPrint(component + " init " + componentHcStr + ", vvmode is " + reconf_vvmode + 
+			  ". Set value as v2 " + reconf_v2);//  + " myConf is " + myConf.hashCode());
+      }
+
+      if (reconf_vvmode.equals("v1v2")) { // reconfiguration injection
+          try {
+              //synchronized(this) {
+                  if (reconf_component.equals(component)) {
+                      if (reconf_point_int == -1) { //FF_ODD
+                          reconf_init_point_index ++;
+                          if ((reconf_init_point_index % 2) == 1) {
+                              myConf.set(reconf_parameter, reconf_v2);
+                              myPrint(component + " init " + componentHcStr + ", PERFORM V1V2 FF_ODD RECONF " + reconf_point +
+					      ". Set value as v2 " + reconf_v2);// + " myConf is " + myConf.hashCode());
+                          } else {
+                              myConf.set(reconf_parameter, reconf_v1);
+                              myPrint(component + " init " + componentHcStr + ", irrelevant init point " + reconf_init_point_index +
+					      ". Set value as v1 " + reconf_v1);// + " myConf is " + myConf.hashCode());
+                          }
+                      } else if (reconf_point_int == -2) { //FF_EVEN
+                          reconf_init_point_index ++;
+                          if ((reconf_init_point_index % 2) == 0) {
+                              myConf.set(reconf_parameter, reconf_v2);
+                              myPrint(component + " init " + componentHcStr + ", PERFORM V1V2 FF_EVEN RECONF " + reconf_point +
+					      ". Set value as v2 " + reconf_v2);// + " myConf is " + myConf.hashCode());
+                          } else {
+                              myConf.set(reconf_parameter, reconf_v1);
+                              myPrint(component + " init " + componentHcStr + ", irrelevant init point " + reconf_init_point_index +
+					      ". Set value as v1 " + reconf_v1);// + " myConf is " + myConf.hashCode());
+                          }
+		      } else {
+                          reconf_init_point_index ++;
+                          if (reconf_point_int == reconf_init_point_index) {
+                              myConf.set(reconf_parameter, reconf_v2);
+                              myPrint(component + " init " + componentHcStr + ", PERFORM V1V2 RECONF " + reconf_point +
+					      ". Set value as v2 " + reconf_v2);// + " myConf is " + myConf.hashCode());
+                          } else {
+                              myConf.set(reconf_parameter, reconf_v1);
+                              myPrint(component + " init " + componentHcStr + ", irrelevant init point " + reconf_init_point_index +
+					      " not " + reconf_point +
+					      ". Set value as v1 " + reconf_v1);// + " myConf is " + myConf.hashCode());
+                          }
+                      }
+                  } else { // for other component instances, just configure it to be v1
+                      myConf.set(reconf_parameter, reconf_v1);
+                      myPrint(component + " init " + componentHcStr + ", irrelevant component." +
+				      " Set value as v1 " + reconf_v1);// + " myConf is " + myConf.hashCode()); 
+                  }
+             // }
+          } catch (Exception e) {
+              myPrint("ERROR happened during performReconf");
+              System.exit(1);
+          }
+      }
+    }
+
+    private static void myPrint(String str) { System.out.println("msx-reconfagent " + str);}
+    
+    /*public ReconfAgent() {
+	String callerClassName = new Exception().getStackTrace()[1].getClassName();
+ 	myPrint("creating ReconfAgent " + this.hashCode() + " for " + callerClassName);
+    }*/
+    
+    /*public void checkReconfAtShutdown(Object obj, Configuration myConf) {
+	this.checkReconfAtShutdown(obj, obj.getClass().getSimpleName(), myConf);
+    }*/
+
+    /*public void performReconf(Object obj, Configuration myConf) {
+	this.performReconf(obj, obj.getClass().getSimpleName(), myConf);
+    }*/
+}
diff -ruN /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/qjournal/server/JournalNode.java ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/qjournal/server/JournalNode.java
--- /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/qjournal/server/JournalNode.java	2018-05-05 10:34:35.000000000 +0000
+++ ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/qjournal/server/JournalNode.java	2020-06-03 22:30:26.305074214 +0000
@@ -106,6 +106,8 @@
 
   @Override
   public void setConf(Configuration conf) {
+    // msx
+    org.apache.hadoop.conf.Configured.performReconf(this, "hdfs:JournalNode", conf);
     this.conf = conf;
     this.localDir = new File(
         conf.get(DFSConfigKeys.DFS_JOURNALNODE_EDITS_DIR_KEY,
diff -ruN /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java
--- /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java	2018-05-05 10:34:35.000000000 +0000
+++ ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java	2020-06-03 22:30:26.297074122 +0000
@@ -401,6 +401,8 @@
   @InterfaceAudience.LimitedPrivate("HDFS")
   DataNode(final Configuration conf) {
     super(conf);
+    // msx
+    org.apache.hadoop.conf.Configured.performReconf(this, "hdfs:DataNode", conf);
     this.blockScanner = new BlockScanner(this, conf);
     this.tracer = createTracer(conf);
     this.tracerConfigurationManager =
@@ -425,6 +427,8 @@
            final List<StorageLocation> dataDirs,
            final SecureResources resources) throws IOException {
     super(conf);
+    // msx
+    org.apache.hadoop.conf.Configured.performReconf(this, "hdfs:DataNode", conf);
     this.tracer = createTracer(conf);
     this.tracerConfigurationManager =
         new TracerConfigurationManager(DATANODE_HTRACE_PREFIX, conf);
diff -ruN /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java
--- /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java	2018-05-05 10:34:35.000000000 +0000
+++ ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java	2020-06-03 22:30:26.253073620 +0000
@@ -887,6 +887,9 @@
 
   protected NameNode(Configuration conf, NamenodeRole role)
       throws IOException {
+    // msx
+    org.apache.hadoop.conf.Configured.performReconf(this, "hdfs:NameNode", conf);
+
     this.tracer = new Tracer.Builder("NameNode").
         conf(TraceUtils.wrapHadoopConf(NAMENODE_HTRACE_PREFIX, conf)).
         build();
@@ -1564,6 +1567,8 @@
 
     switch (startOpt) {
       case FORMAT: {
+        // msx
+        org.apache.hadoop.conf.Configured.performReconf(null, "hdfs:NameNode_FORMAT", conf);
         boolean aborted = format(conf, startOpt.getForceFormat(),
             startOpt.getInteractiveFormat());
         terminate(aborted ? 1 : 0);
@@ -1583,17 +1588,23 @@
         return null; // avoid javac warning
       }
       case ROLLBACK: {
+        // msx
+        org.apache.hadoop.conf.Configured.performReconf(null, "hdfs:NameNode_ROLLBACK", conf);
         boolean aborted = doRollback(conf, true);
         terminate(aborted ? 1 : 0);
         return null; // avoid warning
       }
       case BOOTSTRAPSTANDBY: {
+        // msx
+        org.apache.hadoop.conf.Configured.performReconf(null, "hdfs:NameNode_BOOTSTRAPSTANDBY", conf);
         String toolArgs[] = Arrays.copyOfRange(argv, 1, argv.length);
         int rc = BootstrapStandby.run(toolArgs, conf);
         terminate(rc);
         return null; // avoid warning
       }
       case INITIALIZESHAREDEDITS: {
+        // msx
+        org.apache.hadoop.conf.Configured.performReconf(null, "hdfs:NameNode_INITIALIZESHAREDEDITS", conf);
         boolean aborted = initializeSharedEdits(conf,
             startOpt.getForceFormat(),
             startOpt.getInteractiveFormat());
@@ -1604,13 +1615,18 @@
       case CHECKPOINT: {
         NamenodeRole role = startOpt.toNodeRole();
         DefaultMetricsSystem.initialize(role.toString().replace(" ", ""));
+	// msx
         return new BackupNode(conf, role);
       }
       case RECOVER: {
+        // msx
+        org.apache.hadoop.conf.Configured.performReconf(null, "hdfs:NameNode_RECOVER", conf);
         NameNode.doRecovery(startOpt, conf);
         return null;
       }
       case METADATAVERSION: {
+        // msx
+        org.apache.hadoop.conf.Configured.performReconf(null, "hdfs:NameNode_METADATAVERSION", conf);
         printMetadataVersion(conf);
         terminate(0);
         return null; // avoid javac warning
diff -ruN /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/SecondaryNameNode.java ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/SecondaryNameNode.java
--- /hadoop-2.8.5-src/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/SecondaryNameNode.java	2018-05-05 10:34:35.000000000 +0000
+++ ./hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/SecondaryNameNode.java	2020-06-03 22:30:26.261073711 +0000
@@ -183,6 +183,8 @@
   
   public SecondaryNameNode(Configuration conf,
       CommandLineOpts commandLineOpts) throws IOException {
+    // msx
+    org.apache.hadoop.conf.Configured.performReconf(this, "hdfs:SecondaryNameNode", conf);
     try {
       String nsId = DFSUtil.getSecondaryNameServiceId(conf);
       if (HAUtil.isHAEnabled(conf, nsId)) {
