diff -ruN /hadoop-3.2.1-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/service/AbstractService.java ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/service/AbstractService.java
--- /hadoop-3.2.1-src/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/service/AbstractService.java	2019-09-10 14:35:49.000000000 +0000
+++ ./hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/service/AbstractService.java	2020-04-10 05:15:01.791942858 +0000
@@ -149,7 +149,7 @@
    * the state change not permitted, or something else went wrong
    */
   @Override
-  public void init(Configuration conf) {
+  public void init(Configuration conf) { 
     if (conf == null) {
       throw new ServiceStateException("Cannot initialize service "
                                       + getName() + ": null configuration");
diff -ruN /hadoop-3.2.1-src/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/MiniYARNCluster.java ./hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/MiniYARNCluster.java
--- /hadoop-3.2.1-src/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/MiniYARNCluster.java	2019-09-10 14:35:50.000000000 +0000
+++ ./hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-tests/src/test/java/org/apache/hadoop/yarn/server/MiniYARNCluster.java	2020-04-10 05:24:32.063807941 +0000
@@ -496,10 +496,12 @@
         throws Exception {
       initResourceManager(index, conf);
       super.serviceInit(conf);
+      LOG.info("msx-yarn ResourceManager init");
     }
 
     @Override
     protected synchronized void serviceStart() throws Exception {
+      LOG.info("msx-yarn ResourceManager start");
       startResourceManager(index);
       if(index == 0 && resourceManagers[index].getRMContext().isHAEnabled()) {
         resourceManagers[index].getRMContext().getRMAdminService()
@@ -529,6 +531,7 @@
     
     @Override
     protected synchronized void serviceStop() throws Exception {
+      LOG.info("msx-yarn ResourceManager stop");
       if (resourceManagers[index] != null) {
         waitForAppMastersToFinish(5000);
         resourceManagers[index].stop();
@@ -599,6 +602,7 @@
       LOG.info("Starting NM: " + index);
       nodeManagers[index].init(config);
       super.serviceInit(config);
+      LOG.info("msx-yarn NodeManager init");
     }
 
     /**
@@ -622,6 +626,7 @@
     }
 
     protected synchronized void serviceStart() throws Exception {
+      LOG.info("msx-yarn NodeManager start");
       nodeManagers[index].start();
       if (nodeManagers[index].getServiceState() != STATE.STARTED) {
         // NM could have failed.
@@ -632,6 +637,7 @@
 
     @Override
     protected synchronized void serviceStop() throws Exception {
+      LOG.info("msx-yarn NodeManager stop");
       if (nodeManagers[index] != null) {
         nodeManagers[index].stop();
       }
@@ -820,10 +826,12 @@
           MemoryTimelineStateStore.class, TimelineStateStore.class);
       appHistoryServer.init(conf);
       super.serviceInit(conf);
+      LOG.info("msx-yarn ApplicationHistoryServer init");
     }
 
     @Override
     protected synchronized void serviceStart() throws Exception {
+      LOG.info("msx-yarn ApplicationHistoryServer start");
       appHistoryServer.start();
       if (appHistoryServer.getServiceState() != STATE.STARTED) {
         // AHS could have failed.
@@ -842,6 +850,7 @@
 
     @Override
     protected synchronized void serviceStop() throws Exception {
+      LOG.info("msx-yarn ApplicationHistoryServer stop");
       if (appHistoryServer != null) {
         appHistoryServer.stop();
       }
